### ISSUE TITLE
Not working with TypeScript v5 (author's note)

### ISSUE THREAD

**Issue by nonara** [nonara/ts-patch#93]
# Update 

Beta version is out! See the message below for detail:

- https://github.com/nonara/ts-patch/issues/93#issuecomment-1500795063

**Comment by moshest** [comment]
Thank you for all the effort you put on this.

I just want to raise a point that waiting until the end of April can be too much of a delay for some projects (mine in particular: https://github.com/neuledge/engine-js). Given the fact that the beta was released almost 6 months ago the current situation isn't promising and it may push users to seek for other solutions (including myself).

There is any way we can speed up the migration? 🙏
Can we at least share what's the progress being done and what's still missing?

**Comment by nonara** [comment]
A few points...

1. This is FOSS. In the years this and my other large libraries have been active, I believe $75 in total has been donated.

2. I run a company, work 6 days per week, and generally work over 40hrs per week.

3. I will be surrendering revenue and time to get this done. As of the last few months, I simply and literally could not do this, for reasons which are personal.

I say this with personal respect, but also with the due response to such a message - if you'd like to take your "business" elsewhere, you're more than welcome to do so. However, at present, this is a problem that is not yet solved and is, as I've said, non-trivial.

**Comment by nonara** [comment]
A few points...

1. This is FOSS. In the years this and my other large libraries have been active, I believe $75 in total has been donated.

2. I run a company, work 6 days per week, and generally work over 40hrs per week.

3. I will be surrendering revenue and time to get this done. As of the last few months, I simply and literally could not do this, for reasons which are personal.

I say this with personal respect, but also with the due response to such a message - if you'd like to take your "business" elsewhere, you're more than welcome to do so. However, at present, this is a problem that is not yet solved and is, as I've said, non-trivial.

Lastly, I also think this is important to point out -

If this was an issue that prevented current libraries from working, that is one  thing, but rewriting everything to support a new major version is another. The latter is nice to have early, but the worst case scenario is you wait an additional month to use the latest features.

Demanding a maintainer work more quickly for free so that you can use the newest features is precisely the sort of thing that I've seen embitter and contribute to the burnout of so many formerly enthusiastic maintainers. I say this in the hope that you will more carefully consider your tone and demands in the future in this regard.

**Comment by moshest** [comment]
Hi @nonara,

Thank you for the detailed response. As an open-source contributor and library maintainer I know exactly how you feel.

I never meant to ask you to "work faster" and I apologize for the misunderstanding. Please note I asked how can "we" release it faster. 

My main feedback is that it's not clear what needed to be done in order to support version 5, nor a clear tasks or milestones  that can help a community member such myself to hop on and contribute a few hours on the weekend.

**Comment by nonara** [comment]
@moshest I appreciate that. Neuledge looks nice, btw!

Unfortunately, this may be a complicated problem. I believe I have an approach worked out, though it's a bit in depth as well.

There's a chance we can do it more simply, similar to how we have been. The MSFT engineer who made the changes [isn't sure it will be that easy](https://github.com/nonara/ts-patch/issues/83), though the more I think about it, the more I think it may be.

One quick thing that would help to investigate that angle — if you want to see if you can get TS v5 to build without tree shaking and send me the compiled `typescript.js` and `tsc.js` files, I will be able to take a look and see if the simple approach will work.

I do have one other person who's interested in helping with support and the rewrite. If you're interested, I can send a slack invite.

**Comment by AlCalzone** [comment]
> If you're interested, I can send a slack invite.

You can send me one. I'm relying heavily on `ts-patch` and should be able to spare some time to help out.

**Comment by moshest** [comment]
@nonara I think I manage to build Typescript without tree shaking but the differences aren't big so maybe I only manage to do so partially:
- `tsc.js` - 5.7MB instead of 5.8MB
- `typescript.js` - same size 8.2MB

What I did is update the file `Herebyfile.mjs`:

```diff
+ treeShaking: false,
- treeShaking: taskOptions.treeShaking,
```

I hope it helps.

**Comment by nonara** [comment]
@moshest Thank you. I sent you a slack invite if you'd like to join. @AlCalzone built it with shaking disabled and sent the files on slack.

Good news — I managed to get it to augment the new tsc. It runs properly and seem to be good. We need to test to be sure. Next steps are easier. Feel free to accept the invite and join in.

**Comment by sunrabbit123** [comment]
How long until ts-patch supports V5? If it's still a ways off, I would like to contribute as well.

Could you share the code blocks for the tests that need to pass and the work that needs to be done?

In ttsc, we hacked createProgram.

**Comment by nonara** [comment]
> How long until ts-patch supports V5? If it's still a ways off, I would like to contribute as well.

Thank you. It's almost there. We've overcome the large hurdle. Will update in this thread shortly.

**Comment by TankOs** [comment]
@nonara I really don't want to bug you, but any info on the release? We're currently thinking about downgrading from TS 5 to 4, or just waiting for a couple of days until your new release hits the road. 🙂

**Comment by nonara** [comment]
@TankOs I appreciate that! I know everyone is ready, and I'm with you on that. I finished everything for the beta last night, and I'm working on some high level e2e tests right now so I can push out a beta. Shooting for today, but it definitely won't be long now.

**Comment by nonara** [comment]
Hi all. We have a beta ready.  🎉 

For now, you can use `yarn add -D ts-patch@v3-beta`. I will keep that tag up to date with the latest beta

- If it works in your project, please leave a 🚀 emoji on this message
- If anything doesn't work, please leave a comment

Please note that this is an early release, intended to get everyone going. Everything works, but I will be doing some more tuning and adding more comprehensive tests over the coming days. 

Thank you for helping to beta test!

## ttypescript migration

If you want to try with a ttypescript project, you can use `tspc` for the compile (instead of `ttsc`) or use `compiler: ts-patch` in a config (instead of `compiler: ttypescript`)

## Helping out

If anyone would like to help build tests, please leave a comment on the [v3-beta branch](https://github.com/nonara/ts-patch/tree/v3-beta). I'm planning to organize how to approach those next week and would appreciate help if anyone would like to.

**Comment by jakebailey** [comment]
Oh wow, to be clear, this branch is taking the same approach as prettier and using the `// src/compiler/...` comments as a region marker to try and apply patches? Hope we never move anything, or stop using esbuild... I almost guaranteet that the order of these modules in the output bundle is going to change in TS 5.1 just through refactors, not to mention that I am hoping to use ESM for our executables (like tsc.js), which will end up looking quite different.

Have you considered taking yarn's approach to patching, and instead fork TS, make the changes in-tree, then diff the output packages to produce a patch? I feel like that would end up safer... and also maybe provide a baseline for a PR into TS itself if we want to eventually accept plugins upstream.

**Comment by nonara** [comment]
@jakebailey Hey Jake! Thanks for taking a look. I'm going to respond in the v3 thread, so we can discuss it there:

- https://github.com/nonara/ts-patch/pull/95#issuecomment-1502280245

**Comment by nonara** [comment]
@jakebailey Hey Jake! Thanks for taking a look. I'm going to respond in the v3 thread, so we can discuss there:

- https://github.com/nonara/ts-patch/pull/96#issuecomment-1502285112

**Comment by samchon** [comment]
When try to run on `ts-node` using `-C ts-patch` argument, such error occurs.

Same error be occured even when loading `ts-node` in JS file like below:

```javascript
const runner = require("ts-node");

runner.register({
    project: __dirname + "/tsconfig.issue.json",
    compiler: "ts-patch",
});
require(`./issues/${process.argv[2]}.ts`);
```

```bash
D:\github\samchon\typia\node_modules\ts-node\dist\configuration.js:91
    const { fileExists = ts.sys.fileExists, readFile = ts.sys.readFile, skipProject = index_1.DEFAULTS.skipProject, project = index_1.DEFAULTS.project, tsTrace = index_1.DEFAULTS.tsTrace, } = rawApiOptions;
                                ^

TypeError: Cannot read properties of undefined (reading 'fileExists')
    at readConfig (D:\github\samchon\typia\node_modules\ts-node\dist\configuration.js:91:33)
    at findAndReadConfig (D:\github\samchon\typia\node_modules\ts-node\dist\configuration.js:50:84)
    at create (D:\github\samchon\typia\node_modules\ts-node\dist\index.js:146:69)
    at Object.register (D:\github\samchon\typia\node_modules\ts-node\dist\index.js:127:19)
    at Object.<anonymous> (D:\github\samchon\typia\test\issue.js:6:8)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
```

**Comment by nonara** [comment]
@samchon Thanks for the report! Two things I'd want to confirm are:

1. Is it using `lib/typescript.js` or another script (ie. tsc) — I'm assuming it's the former
2. Does ts-node work if you specify `{ compiler: 'typescript' }`?

I will have a look when I can in a few days. If you're interested looking into it, it sounds like `ts` exists and is exported properly, but, for whatever reason, module.exports.ts doesn't contain `sys`. 

Possibilities that come to mind:

1. Could be a change to ts5, where sys isn't exported (seems unlikely)
2. If it's trying to use `tsc.js`, I don't think there is a module.exports set
3. If it's using `lib/typescript.js`, there might be something wrong with the runInThisContext binding for exports.

It's probably 3, so if you'd like to debug it, that's where I'd start.

**Comment by samchon** [comment]
https://github.com/samchon/ts-patch-bug-report-93

```bash
git clone https://github.com/samchon/ts-patch-bug-report-93
cd ts-patch-bug-report-93
npm install
npm run build
npm run start
```

@nonara, you can reproduce the bug from above repo and commands.

**Comment by hi-zhaolei** [comment]
@nonara Hi. I just now replaced ttypescript with this beta version to solve the previous problem. It works. but see another error

![image](https://user-images.githubusercontent.com/5736062/231751077-c3288d8e-bc8d-4d80-8574-2f9c2db90f4c.png)

it seems a syntax error.

**Comment by nonara** [comment]
@hi-zhaolei Thanks for the report. We just discovered the same issue. That happens with lower Node versions, like Node 14. We'll be publishing a fix in a few hours.

**Comment by yf-yang** [comment]
Hello, I am using typescript@5.0.4 with ts-patch@v3-beta, but the transformer `typescript-transform-paths` is not working correctly. Not sure which part fails, but I see you are the maintainer of both projects. Would you please take a look? Thanks 😄 

https://github.com/yf-yang/ts-patch-not-working

**Comment by yf-yang** [comment]
> Hello, I am using typescript@5.0.4 with ts-patch@v3-beta, but the transformer `typescript-transform-paths` is not working correctly. Not sure which part fails, but I see you are the maintainer of both projects. Would you please take a look? Thanks 😄
> 
> https://github.com/yf-yang/ts-patch-not-working

I'm sorry, that is my own fault, I didn't add `"afterDeclarations": true`. Thanks for your attention!

**Comment by samchon** [comment]
@nonara When can I download v3 pubilsh? If there're som problem in `tspc`, I think it would better to publish only `ts-patch install` first and `tspc` later.

**Comment by MeirionHughes** [comment]
Same happens when using env ```cross-env TS_NODE_COMPILER=ts-patch ...``` as well as ```ts-node --compiler ts-patch ...```

**Comment by robert-w-gries** [comment]
Hi @nonara it appears I got booted out of the Slack channel due to not having Slack pro (it's unclear why I needed Pro to stay in the channel), so I'd like to check in and see what the v3 changes will entail.

I saw your comment [here](https://github.com/nonara/ts-patch/pull/96#issuecomment-1502285112) and was interested in this part in particular:

> The short version is, you'll be able to bundle language service plugins and transformers in addition to two new features, which are middleware "hooks" and compiler transformers. The latter are applied at patch time on the compiler code itself. This should open up some unique opportunities, which I'm excited to explore.

Is there anything in v3 that I can experiment with in terms of these middleware hooks? Or is that a future, planned feature?

Either way I'll test the v3 branch, hopefully tomorrow, to help out with the release. Thanks for all your work!

**Comment by robert-w-gries** [comment]
Hi @nonara it appears I got booted out of the Slack channel due to not having Slack pro (it's unclear why I needed Pro to stay in the channel), so I'd like to check in and see what the v3 changes will entail.

I saw your comment [here](https://github.com/nonara/ts-patch/pull/96#issuecomment-1502285112) and was interested in this part in particular:

> The short version is, you'll be able to bundle language service plugins and transformers in addition to two new features, which are middleware "hooks" and compiler transformers. The latter are applied at patch time on the compiler code itself. This should open up some unique opportunities, which I'm excited to explore.

Is there anything in v3 that I can experiment with in terms of these middleware hooks? Or is that a future, planned feature?

Either way I'll test the v3 branch, hopefully tomorrow, to help out with the release. Thanks for all your work!

**Comment by frankie303** [comment]
for those who try to install it: `yarn add -D ts-patch@v3-beta` --> `yarn add -D ts-patch@3.0.0-beta3`

**Comment by AlCalzone** [comment]
I've looked at this and the problem seems to be the following:
- `ttypescript` has `lib/typescript.js` as its `main` field, so that gets loaded when specifying it as the compiler.
- `ts-patch` has its own `index.js` as the `main` field, which makes it so it can be used as a library, but not directly as the compiler.

I see two options:
1. change the main field of `ts-patch`, so it points to the patched `typescript.js`, and expose the library functionality otherwise.
2. add subpath `exports` to the `package.json`, e.g. `./compiler`, which points to `lib/typescript.js`. That way `ts-node -c ts-patch/compiler` would resolve the correct file.

**Comment by nonara** [comment]
Ok! Finally was able to wrap things up over the last few days.

@khugin Thank you for investigating the issue. You were correct. I assumed that the logic behind the "compiler" option would generally be to look for "lib/typescript.js", but I didn't recall that typescript's `package.json` `main` entry was pointed to it.

# New Changes

I've changed the `lib` directory to `compiler`. So you can specify `ts-patch/compiler` for tools like ts-node, or anything that lets you specify a compiler.

I also built some tests which cover the e2e well enough for now.

Aside from that, I finished some other major features that should close all open issues, including:

1. Mutexes (to prevent cascading issues when multiple times in parallel)
2. ESM support for transformers (this ended up being as not fun as you can imagine, but it works!)
3. Uses native compiler API, so you no longer need ts-node or tsconfig-paths _(let me know if you have any issues)_
4. Various edge case issues fixed

Some of these (especially 2 & 3) might trigger some edge cases from the last beta, so feel free to report if so.

I've gone ahead and updated the `v3-beta` tag for now. Alternatively, you can directly use `3.0.0-rc`.

Our tests cover the new things pretty well, and given the good reports from the last, I'm going to do a release in a few days, but anyone willing to test it before then, I'd be grateful! 

Go ahead and leave a 🚀 if it works!

**Comment by nonara** [comment]
@robert-w-gries 

> Hi @nonara it appears I got booted out of the Slack channel due to not having Slack pro (it's unclear why I needed Pro to stay in the channel), so I'd like to check in and see what the v3 changes will entail.

Ah, I thought you just left. Sorry about that! It's most likely because I sent it as a Slack Connect invite by mistake. If you're still interested in helping out, let me know, and I will send a proper guest invite.

> Is there anything in v3 that I can experiment with in terms of these middleware hooks? Or is that a future, planned feature?

The foundation is all there. It's a relatively short distance to implementing the middleware hooks. It mainly needs some reorganizing and the API to add custom compiler transformers. That's planned for the next major version.

**Comment by samchon** [comment]
@nonara Not working.

There had not been any problem in `ts-patch@2.1.0`.

You can test in below repo:

  - https://github.com/samchon/typia/tree/features/ts5 -> Not working
  - https://github.com/samchon/typia/tree/features/ts4 -> Working

```bash
git clone https://github.com/samchon/typia -b features/ts5
cd typia
npm install
npm run build:test
```

## TypeScript v4.9
```bash
$ npx ts-patch install
[!] Some files can't be patched! You can run again with --verbose to get specific error detail. The following files are unable to be patched:
  - D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js
  - D:\github\samchon\typia@master\node_modules\typescript\lib\typescript.js

$ npx ts-patch install --verbose
[~] Patching tsc.js in D:\github\samchon\typia@master\node_modules\typescript\lib
[~] Writing backup cache to D:\github\samchon\typia@master\node_modules\.cache\ts-patch\da30b49ec053846f15152e30247a54ab\tsc.js
[!] Cannot patch TS version <5
[~] Patching typescript.js in D:\github\samchon\typia@master\node_modules\typescript\lib
[~] Writing backup cache to D:\github\samchon\typia@master\node_modules\.cache\ts-patch\84f1848a4a4eba19e6d8c1694d4ff746\typescript.js
[~] Writing backup cache to D:\github\samchon\typia@master\node_modules\.cache\ts-patch\84f1848a4a4eba19e6d8c1694d4ff746\typescript.d.ts
[!] Cannot patch TS version <5
[!] Some files can't be patched! You can run again with --verbose to get specific error detail. The following files are unable to be patched:
  - D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js
  - D:\github\samchon\typia@master\node_modules\typescript\lib\typescript.js
```

## TypeScript v5.0
```bash
$ npm run build:test

> typia@3.8.0-dev.20230413 build:test
> rimraf bin && tsc -p test/tsconfig.json

D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:368
                            throw error;
                            ^

Error: Cannot find module './transformers/FileTransformer'
Require stack:
- D:\github\samchon\typia@master\src\transform.ts
- D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js
- D:\github\samchon\typia@master\node_modules\typescript\bin\tsc
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1026:15)
    at Module.require (D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:353:45)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (D:\github\samchon\typia@master\src\transform.ts:8:27)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Module.requireCjs (D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:375:35)
    at Module.require (D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:361:75)
    at require (node:internal/modules/cjs/helpers:108:18)      
    at Object.resolveFactory (D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:310:38)
    at PluginCreator.createTransformers (D:\github\samchon\typia@master\node_modules\typescript\lib\tsc.js:170:41) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'D:\\github\\samchon\\typia@master\\src\\transform.ts',    
    'D:\\github\\samchon\\typia@master\\node_modules\\typescript\\lib\\tsc.js',
    'D:\\github\\samchon\\typia@master\\node_modules\\typescript\\bin\\tsc'
  ]
}
```

**Comment by CarsonF** [comment]
Thanks for all your effort here @nonara.

A small nit pick for me. I'm using Yarn PnP. I have typescript "unplugged", so ts-patch can modify it. 
It would be nice if `node_modules/.cache` wasn't assumed. Currently this is the only process that assumes that. I know it's possible to use the PnP API to grab the typescript folder, so maybe that would be a better mutable cache location.
<details>
<summary>FS tree screenshot</summary>


![Screenshot 2023-05-02 at 8 44 43 AM](https://user-images.githubusercontent.com/932566/235685090-84238408-87c0-4a19-9e0b-fe6f1c36e926.png)


</details>

**Comment by nonara** [comment]
@CarsonF The cache dir is actually an established convention, which is why we rely on it: https://www.npmjs.com/package/find-cache-dir

However, in anticipation of edge cases, I added the `TSP_CACHE_DIR` env var. That will go into the new docs when I release.

**Comment by CarsonF** [comment]
Of course sindresorhus has a package for this lol. Looks like that also checks for `CACHE_DIR` env var, so I can use that. Thanks for the info!

**Comment by CarsonF** [comment]
If it helps here's how to get the typescript path with the PnP API.
```ts
pnp = require('pnpapi');
versionReference = pnp.getPackageInformation(pnp.topLevel).packageDependencies.get('typescript');
pnp.getPackageInformation(pnp.getLocator('typescript', versionReference)).packageLocation;
```
```
/.../.yarn/unplugged/typescript-patch-38a6667774/node_modules/typescript/
```

**Comment by donflopez** [comment]
It'd be awesome if ts directly supported this feature or facilitated this libs work, have you seen this?

https://github.com/microsoft/TypeScript/issues/52953

**Comment by mindplay-dk** [comment]
> It'd be awesome if ts directly supported this feature or facilitated this libs work, have you seen this?
> 
> [microsoft/TypeScript#52953](https://github.com/microsoft/TypeScript/issues/52953)

Someone is now taking the initiative and pushing for this feature:

https://github.com/microsoft/TypeScript/pull/54278

@nonara would you care to stop by and see what they're doing? 🙂

**Comment by samchon** [comment]
@mindplay-dk Unfortunately, transform API of TS 5.2 does not support full spec of TypeScript.

For an example, you can't get detailed type info, due to TS 5.2 transform API does not support it.

**Comment by jakebailey** [comment]
I am utterly confused by your description of the situation. Yes, tsc has fewer helpers included, but it's preposterous to state that tsc "does not support the full spec of TypeScript".

Your continued misrepresentation of the way TypeScript works is extremely frustrating; I don't even know why I'm even bothering.

**Comment by samchon** [comment]
> @samchon I am utterly confused by your description of the situation. Yes, tsc has fewer helpers included, but it's preposterous to state that tsc "does not support the full spec of TypeScript".
> 
> Your continued misrepresentation of the way TypeScript works is extremely frustrating; I don't even know why I'm even bothering.

https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551366861

This is my answer. Sorry for bothering you, but it is a break and critical change for 3rd party libraries.

**Comment by sinclairzx81** [comment]
@samchon This [workbench](https://sinclairzx81.github.io/typebox-workbench/?share=LyoqCiAqIEBtaW5MZW5ndGggMTAKICovCnR5cGUgRm9vID0gc3RyaW5nCgpleHBvcnQgaW50ZXJmYWNlIFZlY3RvciB7CiAgLyoqCiAgICogQG1pbmltdW0gMAogICAqIEBtYXhpbXVtIDEKICAgKiBAZGVmYXVsdCAwCiAgICovCiAgeDogbnVtYmVyCiAgLyoqCiAgICogQG1pbmltdW0gMAogICAqIEBtYXhpbXVtIDEKICAgKiBAZGVmYXVsdCAxCiAgICovCiAgeTogbnVtYmVyCn0%3D) thing is implemented without any 3rd party libraries at all. Of the functions you've listed on https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551366861 as missing, most of them are provided as default exports on the compiler API (and there's enough there to generate runtime assertion logic)

**Comment by samchon** [comment]
> @samchon This [workbench](https://sinclairzx81.github.io/typebox-workbench/?share=LyoqCiAqIEBtaW5MZW5ndGggMTAKICovCnR5cGUgRm9vID0gc3RyaW5nCgpleHBvcnQgaW50ZXJmYWNlIFZlY3RvciB7CiAgLyoqCiAgICogQG1pbmltdW0gMAogICAqIEBtYXhpbXVtIDEKICAgKiBAZGVmYXVsdCAwCiAgICovCiAgeDogbnVtYmVyCiAgLyoqCiAgICogQG1pbmltdW0gMAogICAqIEBtYXhpbXVtIDEKICAgKiBAZGVmYXVsdCAxCiAgICovCiAgeTogbnVtYmVyCn0%3D) thing is implemented without any 3rd party libraries at all. Of the functions you've listed on [microsoft/TypeScript#54276 (comment)](https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551366861) as missing, most of them are provided as default exports on the compiler API (and there's enough there to generate runtime assertion logic)

Thanks for reference. It is time to study.

**Comment by nonara** [comment]
Hi all,

To answer the question — I did see the proposal. You can see my responses on that thread, starting with this one:

- https://github.com/microsoft/TypeScript/issues/54276#issuecomment-1551887816

**Comment by mgabeler-lee-6rs** [comment]
Running into this error trying to convert from ttypescript to ts-patch@3.0.0-rc2 (or beta3, same error either way):
```
evalmachine.<anonymous>:370
                            throw error;
                            ^

RangeError: Maximum call stack size exceeded
    at Object.<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:10:23)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Module.requireCjs (evalmachine.<anonymous>:377:35)
    at Module.require (evalmachine.<anonymous>:363:75)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8:17)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Module.requireCjs (evalmachine.<anonymous>:377:35)
    at Module.require (evalmachine.<anonymous>:363:75)
    at require (node:internal/modules/cjs/helpers:108:18)
```
This sadly is a re-throw losing the original stack trace. After some trial & error, I was able to observer more or less what the underlying call stack is. This is from before it crashes, but you can see the loop:
<details>
<summary>nasty long stack trace</summary>

```
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:10)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/asyncify.js:16)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/wrapAsync.js:8)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/internal/eachOfLimit.js:19)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/eachOfLimit.js:7)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/eachOf.js:15)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/async@3.2.4/node_modules/async/forEach.js:7)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/winston@3.7.2/node_modules/winston/lib/winston/logger.js:11)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/winston@3.7.2/node_modules/winston/lib/winston/create-logger.js:12)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/winston@3.7.2/node_modules/winston/lib/winston.js:42)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
7773 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
8674 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
5090 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
7124 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
5831 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
5250 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
2698 (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
r (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/ts-auto-mock@3.6.4_typescript@5.0.4/node_modules/ts-auto-mock/transformer/index.js:1)
Module._compile (internal/modules/cjs/loader:1196)
requireCjs (<node_internals>/evalmachine.<anonymous>:377)
Module.require (<node_internals>/evalmachine.<anonymous>:363)
require (internal/modules/cjs/helpers:108)
resolveFactory (<node_internals>/evalmachine.<anonymous>:312)
createTransformers (<node_internals>/evalmachine.<anonymous>:172)
newEmit (<node_internals>/evalmachine.<anonymous>:102)
computeDtsSignature (<node_internals>/evalmachine.<anonymous>:119455)
updateShapeSignature (<node_internals>/evalmachine.<anonymous>:119485)
getFilesAffectedByWithOldState (<node_internals>/evalmachine.<anonymous>:119443)
getNextAffectedFile (<node_internals>/evalmachine.<anonymous>:119897)
getSemanticDiagnosticsOfNextAffectedFile (<node_internals>/evalmachine.<anonymous>:120653)
getSemanticDiagnostics (<node_internals>/evalmachine.<anonymous>:120689)
emitFilesAndReportErrors (<node_internals>/evalmachine.<anonymous>:122075)
emitFilesAndReportErrorsAndGetExitStatus (<node_internals>/evalmachine.<anonymous>:122106)
performIncrementalCompilation (<node_internals>/evalmachine.<anonymous>:122262)
performIncrementalCompilation2 (<node_internals>/evalmachine.<anonymous>:170589)
executeCommandLineWorker (<node_internals>/evalmachine.<anonymous>:170366)
executeCommandLine (<node_internals>/evalmachine.<anonymous>:170456)
tsp.execTsc (<node_internals>/evalmachine.<anonymous>:170978)
<anonymous> (<node_internals>/evalmachine.<anonymous>:170982)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/ts-patch@3.0.0-rc2/node_modules/ts-patch/compiler/tsc.js:19)
Module._compile (internal/modules/cjs/loader:1196)
Module._extensions..js (internal/modules/cjs/loader:1250)
Module.load (internal/modules/cjs/loader:1074)
Module._load (internal/modules/cjs/loader:909)
Module.require (internal/modules/cjs/loader:1098)
require (internal/modules/cjs/helpers:108)
<anonymous> (CENSORED/common/temp/node_modules/.pnpm/ts-patch@3.0.0-rc2/node_modules/ts-patch/bin/tspc.js:9)
Module._compile (internal/modules/cjs/loader:1196)
Module._extensions..js (internal/modules/cjs/loader:1250)
Module.load (internal/modules/cjs/loader:1074)
Module._load (internal/modules/cjs/loader:909)
executeUserEntryPoint (internal/modules/run_main:81)
<anonymous> (internal/main/run_main_module:22)
```
</details>

The trace above is from my "real" app, but I can reproduce the same problem with this minimal setup:

<details>
<summary>reproducer files</summary>

tsconfig.json:
```json
{
  "compilerOptions": {
    "allowUnreachableCode": false,
    "declaration": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noImplicitReturns": true,
    "preserveConstEnums": true,
    "removeComments": false,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "sourceMap": true,
    "declarationMap": true,
    "strict": true,
    "target": "es2021",
    "types": ["node"],
    "lib": ["es2021"],

    "rootDir": ".",
    "outDir": "dist",
    "plugins": [{ "transform": "ts-auto-mock/transformer" }]
  }
}
```
package.json:
```json
{
  "dependencies": {
    "@types/async": "^3.2.20",
    "@types/node": "^14",
    "async": "^3.2.4",
    "ts-auto-mock": "^3.7.0",
    "ts-patch": "3.0.0-rc2",
    "typescript": "~5.0.4"
  }
}
```
index.ts:
```typescript
import { createMock } from 'ts-auto-mock';

interface X {
        x(): void;
}

const mockX = createMock<X>();
```
</details>

**Comment by robert-w-gries** [comment]
Hi, I've finally gotten some time to work on my project using ts-patch, and it looks to work perfectly when running patched `tsc`.

However, I ran into an issue with patched `tsserver` where the transformer runs as expected, but the patched host code does not seem to be executed. I've confirmed that the transformer runs as expected in VS Code by adding a console.log here:

```js
export default function programTransformer(
    program: Program,
    host: CompilerHost | undefined,
    config: PluginConfig,
    { ts: tsInstance }: ProgramTransformerExtras,
  ): Program {
    const compilerOptions = program.getCompilerOptions();
    const compilerHost = getPatchedHost(host, tsInstance, compilerOptions);
    const rootFileNames = program.getRootFileNames().map(tsInstance.normalizePath);
    console.log("CREATING PROGRAM")
    return tsInstance.createProgram(rootFileNames, compilerOptions, compilerHost);
};
```

However, I've added console logs inside the return of the `getPatchedHost` object, and it doesn't seem like the patched host object is used in tsserver.

Here is my settings.json:

```json
  {
    "typescript.tsdk": "./node_modules/typescript/lib",
    "typescript.tsserver.log": "verbose", // to debug, set to "verbose"
    "typescript.enablePromptUseWorkspaceTsdk": true
  }
```

I have verified the patch is installed for all js files:

```bash
$ ./node_modules/.bin/ts-patch check
Checking TypeScript v5.0.4 installation in $PROJECT\demo\node_modules\typescript

[+] tsc.js is patched with ts-patch version 3.0.0-rc2 
[+] tsserver.js is patched with ts-patch version 3.0.0-rc2
[+] tsserverlibrary.js is patched with ts-patch version 3.0.0-rc2 
[+] typescript.js is patched with ts-patch version 3.0.0-rc2 
```

If needed, I can add more details about my project's weird use of ts-patch in the slack as I don't want to derail this issue.

**Comment by AlCalzone** [comment]
I have a patch up for the infinite loop: https://github.com/nonara/ts-patch/pull/99

At least for my codebase with already-compiled `.js` transformers, this avoids the infinite cycle, as the default node `require` algorithm seems to be able to deal with a require cycle.

**Comment by jeka2100** [comment]
Hello, I migrating from ttypescript: 

1. install ts-patch@v3-beta
2. Run "tspc --build"

And has error:
<--- Last few GCs --->

[70676:0x7fc848008000]    49489 ms: Scavenge (reduce) 4078.7 (4102.0) -> 4078.5 (4102.5) MB, 2.8 / 0.0 ms  (average mu = 0.253, current mu = 0.145) allocation failure 
[70676:0x7fc848008000]    50861 ms: Mark-sweep (reduce) 4095.1 (4118.2) -> 4092.3 (4118.4) MB, 1251.9 / 0.0 ms  (+ 2.3 ms in 1 steps since start of marking, biggest step 2.3 ms, walltime since start of marking 1260 ms) (average mu = 0.176, current mu = 0.

<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10f2193a5 node::Abort() (.cold.1) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 2: 0x10def4869 node::Abort() [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 3: 0x10def49df node::OnFatalError(char const*, char const*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 4: 0x10e073347 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 5: 0x10e0732e3 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 6: 0x10e2322e5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 7: 0x10e230d06 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 8: 0x10e23d240 v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
 9: 0x10e23d2c1 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
10: 0x10e202b9e v8::internal::FactoryBase<v8::internal::Factory>::NewWeakFixedArrayWithMap(v8::internal::Map, int, v8::internal::AllocationType) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
11: 0x10e133c8c v8::internal::Handle<v8::internal::SharedFunctionInfo> v8::internal::(anonymous namespace)::CreateTopLevelSharedFunctionInfo<v8::internal::Isolate>(v8::internal::ParseInfo*, v8::internal::Handle<v8::internal::Script>, v8::internal::Isolate*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
12: 0x10e13053b v8::internal::(anonymous namespace)::CompileToplevel(v8::internal::ParseInfo*, v8::internal::Handle<v8::internal::Script>, v8::internal::MaybeHandle<v8::internal::ScopeInfo>, v8::internal::Isolate*, v8::internal::IsCompiledScope*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
13: 0x10e132f1a v8::internal::Compiler::GetWrappedFunction(v8::internal::Handle<v8::internal::String>, v8::internal::Handle<v8::internal::FixedArray>, v8::internal::Handle<v8::internal::Context>, v8::internal::Compiler::ScriptDetails const&, v8::ScriptOriginOptions, v8::internal::ScriptData*, v8::ScriptCompiler::CompileOptions, v8::ScriptCompiler::NoCacheReason) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
14: 0x10e07ddf7 v8::ScriptCompiler::CompileFunctionInContext(v8::Local<v8::Context>, v8::ScriptCompiler::Source*, unsigned long, v8::Local<v8::String>*, unsigned long, v8::Local<v8::Object>*, v8::ScriptCompiler::CompileOptions, v8::ScriptCompiler::NoCacheReason, v8::Local<v8::ScriptOrModule>*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
15: 0x10dee695b node::contextify::ContextifyContext::CompileFunction(v8::FunctionCallbackInfo<v8::Value> const&) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
16: 0x10e0db54e v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
17: 0x10e0dac46 v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
18: 0x10e0da39f v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
19: 0x10e95b3d9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit [/Users/user/.nvm/versions/node/v16.9.1/bin/node]
sh: line 1: 70676 Abort trap: 6           tspc

**Comment by nonara** [comment]
## 🎉 **rc3 is ready**

Issues are resolved and all of the new features also appear to be working. Test coverage isn't as broad as it will be, but I'm fairly comfortable at this point.

I'd like to push this through as a full release very quickly, _so testers are appreciated!_

You can pull it via the `v3-beta` tag.
Or by exact version: `3.0.0-rc3`


**If you test it, please leave 🎉  if it works or a comment if it doesn't**

## Post-mortem

If you're technically minded and wondering, this is what was going on:

The heap overflow was simply due to not to integrating NodeJS's module caching. I had incorrectly assumed this would be covered by node, but upon looking closer, I realized that it needed to be manually added. As a result, anything which had a circular require pattern caused an infinite loop.

I also took some time to optimize and improved the require hooking logic.

## Other concerns

Please do leave a comment on any issues you face with the latest release candidate in this thread. I'm going to push a release in a few days after people have had a chance to test.

@robert-w-gries Regarding your issue, please do post some details in Slack. I can take a look.

We can't really guarantee tsserver working the way people would like them to in IDEs, because IDEs tend to do their own magic.  It doesn't sound to me like an issue with tsp, but more of some unknown behaviour in how vscode is using the API.

Regardless, I might be able to help track out what's going on. From the sound of it, you're saying it's making a `Program` but the `CompilerHost` functions don't seem to be working — that definitely seems odd to me, but let's keep that discussion in its own thread.

**Comment by samchon** [comment]
@nonara Works on `beta3`, but fallen into infinite error loop on `rc3`

  - https://github.com/samchon/typia/actions/runs/5087037428/jobs/9142042595#step:7:9

When you clone and try to build my project `typia`, you can see that below error message be repeated infinitey.

Also, until beta3 version (including previous versions like `ts-patch@2.1.0`), it was possible to configuring `ts` transformer file directly, therefore it made trasformer library development conveniently. However, since `rc3` update, using `ts` transformer file became impossible.

```json
{
  "compilerOptions": {
    "plugins": [
      { "transform": "../src/transform.ts" } // possible until beta3, but impossible since rc3
      { "transform": "../lib/trasnform.js" } // therefore, must target to compiled JS fil
    ]
  }
}
```

## Reproduce
```bash
git clone https://github.com/samchon/typia -b ts-patch-3.0.0-rc3
cd typia
npm install
npm run build
npm run build:test
```

## Error Message
```bash
Error: Error on typia.application(): no transform has been configured. Run "npx typia setup" command.
    at halt (/home/runner/work/typia/typia/bin/src/module.js:479:11)
    at Object.application (/home/runner/work/typia/typia/bin/src/module.js:137:5)
    at Object.<anonymous> (/home/runner/work/typia/typia/bin/test/features/application/ajv/test_application_ajv_ArrayAny.js:9:119)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32)
    at Function.Module._load (node:internal/modules/cjs/loader:909:12)
    at Module.require (node:internal/modules/cjs/loader:1098:19)
    at require (node:internal/modules/cjs/helpers:108:[18](https://github.com/samchon/typia/actions/runs/5087037428/jobs/9142042595#step:8:19))
    at /home/runner/work/typia/typia/bin/test/helpers/DynamicImportIterator.js:66:89
```

**Comment by nonara** [comment]
@samchon Thanks for trying it out and your report! The rc does support ts transformers, but your case brought some edge cases to light. 

One of the more ambitious aspects of the new version is replacing need for ts-node, by doing ts compilation natively. Looks like I need to cover more edge cases in the require hook to cover implicit extensions, etc. 

I'm going to look closer at what ts-node is doing for the require hook, and hopefully I'll have it sorted in the next day or two.

**Comment by AlCalzone** [comment]
@nonara I'm getting the same too, but that just looks like you forgot to remove debug logging from the `finally` here:
https://github.com/nonara/ts-patch/blob/8a8cb09e8c83d19f94c9528d5c7388459f811477/projects/patch/src/plugin/plugin-creator.ts#L101-L105

**Comment by TimUnderhay** [comment]
Not sure if this helps, @nonara, but the same error is logged many times with Typescript 5.1.3, with both beta3 or rc3.  Reverting to TS 5.0.4 makes the errors go away.

```Error
    at tspWrappedFactory (evalmachine.<anonymous>:350:33)
    at evalmachine.<anonymous>:110045:33
    at evalmachine.<anonymous>:110132:61
    at Array.map (<anonymous>)
    at transformNodes (evalmachine.<anonymous>:110132:50)
    at emitJsFileOrBundle (evalmachine.<anonymous>:110730:26)
    at emitSourceFileOrBundle (evalmachine.<anonymous>:110665:7)
    at forEachEmittedFile (evalmachine.<anonymous>:110419:26)
    at emitFiles (evalmachine.<anonymous>:110639:5)
    at emitWorker (evalmachine.<anonymous>:117945:30)
```

**Comment by nonara** [comment]
Finally got the time to sit down with this for a few days and wrap it up. Thank you all for your patience and for helping with testing!

If there are any issues, go ahead and file as a new issue.

**Comment by carlossalasamper** [comment]
> Hi all. We have a beta ready that works for both `ts-patch` and `ttypescript` projects! 🔥
> 
> I'd really appreciate you trying it out and letting me know if it works for you.
> 
> For now, you can use `yarn add -D ts-patch@v3-beta`. I will keep that tag up to date with the latest beta
> 
> * If it works in your `ts-patch` based project, please leave a 🚀 emoji on this message
> * If it works in your `ttypescript` based project, please leave a 🎉 emoji
> * If anything doesn't work, please leave a comment
> 
> Please note that this is an early release, intended to get everyone going. Everything should work, but I will be doing some more tuning and adding more comprehensive tests over the coming days.
> 
> Thank you for helping to beta test!
> 
> ## ttypescript migration
> If you want to try with a ttypescript project, you can use:
> 
> * `tspc` to compile (instead of `ttsc`)
> * `{ 'compiler': 'ts-patch/compiler' }` in a config (instead of `{ 'compiler': 'ttypescript' }`)
> 
> Or with ts-node, etc. via cli arg:
> 
> `--compiler ts-patch/compiler`
> 
> ## Helping out
> If anyone would like to help build tests, please leave a comment on the [v3-beta PR](https://github.com/nonara/ts-patch/pull/95). I'm planning to organize how to approach those next week and would appreciate any help available.

Thanks dad
